AEA-521 Aries demo (#1146)
fetchai/agents-aea: docs/aries-cloud-agent-demo.md

<div class="admonition note">
  <p class="admonition-title">Note</p>
  <p>This demo is incomplete and will soon be updated.
</p>
</div>

Demonstrating an entire decentralised identity scenario involving AEAs and instances of Aries Cloud Agents (ACAs).

## Discussion

This demo corresponds with the one <a href="https://github.com/hyperledger/aries-cloudagent-python/blob/master/demo/README.md" target=_blank>here</a> from <a href="https://github.com/hyperledger/aries-cloudagent-python" target=_blank> aries cloud agent repository </a>. 

There are two AEAs: 

 * Alice_AEA
 * Faber_AEA 

and two ACAs:

 * Alice_ACA
 * Faber_ACA
 
Each AEA is connected to its corresponding ACA: Alice_AEA to Alice_ACA and Faber_AEA to Faber_ACA.

The following lists the sequence of interactions between the four agents:

>  * Alice_AEA: starts

Reviewer: A UML diagram would be nice here (see other demos)

Author: Yes. What tool do you use?

 * Alice_AEA: shows its identity in the terminal and waits for an `invitation` detail from Faber_AEA.
 * Faber_AEA: starts
 * Faber_AEA: tests its connection to Faber_ACA.
 * Faber_ACA: responds to Faber_AEA.
 * Faber_AEA: requests Faber_ACA to create an invitation.
 * Faber_ACA: responds by sending back the `connection` detail, which contains an `invitation` field.
 * Faber_AEA: sends the `invitation` detail to Alice_AEA.
 * Alice_AEA: receives `invitation` detail from Faber_AEA.
 * Alice_AEA: requests Alice_ACA to accept the invitation, by passing it the `invitation` detail it received in the last step.

All messages from an AEA to an ACA are http requests (using `http_client` connection).

> All messages from an AEA to another AEA are via the `oef` connection.

Reviewer: ```suggestion
All messages from an AEA to another AEA utilise the `oef` communication network accessed via the `oef` connection.
```

```

Make sure the above ports are unused. To learn more about the above command for starting an aca and its various options: 

``` bash
aca-py start --help
```

Take note of the specific IP addresses and ports you used in the above command. We will refer to them by the following names: 

* Faber admin IP: 127.0.0.1
* Faber admin port: 8021
* Faber webhook port: 8022

The admin IP and port will be used to send administrative commands to this ACA from an AEA. 

The webhook port is where the ACA will send notifications to. We will expose this from the AEA so it receives this ACA's notifications.

### Run Alice_ACA

Type this in the second terminal:

``` bash
aca-py start --admin 127.0.0.1 8031 --admin-insecure-mode --inbound-transport http 0.0.0.0 8030 --outbound-transp http --webhook-url http://127.0.0.1:8032/webhooks
```

Again, make sure the above ports are unused and take note of the specific IP addresses and ports. In this case:

* Alice admin IP: 127.0.0.1
* Alice admin port: 8031
* Alice webhook port: 8032

## Create Alice and Faber AEAs

### Create Alice_AEA

In the third terminal, create an Alice_AEA and move into its project folder: 

``` bash
aea create alice

> cd faber

Reviewer: ```suggestion
cd alice
```

```

### Add and Configure the Skill

Add the `aries_alice` skill:

``` bash

> aea add --local skill fetchai/aries_alice:0.1.0

Reviewer: the guide will be for externals using the registry, so 
```suggestion
aea add skill fetchai/aries_alice:0.1.0
```

```

> You then need to configure this skill. Open the skill's configuration file in `alice/vendor/fetchai/skills/aries_alice/skill.yaml` and ensure `admin_host` and `admin_port` details match those you noted above for Faber_ACA. 

Reviewer: They should use the command line, much easier: `aea config set vendor.fetchai.skills.aries_alice.admin_host ...` etc

### Add and Configure the Connections

Add `http_client`, `oef` and `webhook` connections:

``` bash

> aea add --local connection fetchai/http_client:0.1.0

Reviewer: drop `--local`

aea add --local connection fetchai/webhook:0.1.0

> aea add connection fetchai/oef:0.1.0

Reviewer: its 0.2.0 now for oef

```

> You now need to configure the `webhook` connection. Open the connection's configuration file in `alice/vendor/fetchai/connections/webhook/connection.yaml` and ensure `webhook_port` matches with what you used above for Faber_ACA. 

Reviewer: see above

Next, make sure that the value of `webhook_url_path` is `/webhooks/topic/{topic}/`.

### Configure Alice_AEA:

> You now need to ensure that Alice_AEA uses the OEF connection as its default connection. Open the agent's configuration file in `alice/aea-config.yaml` and ensure that the `default_connection`'s value is `fetchai/oef:0.1.0`.

Reviewer: again, use command line instead

### Install the Dependencies and Run Alice_AEA:

Install the dependencies:

``` bash
aea install
```

Then run Alice_AEA:

``` bash

> aea run --connections fetchai/http_client:0.1.0,fetchai/oef:0.1.0,fetchai/webhook:0.1.0

Reviewer: oef:0.2.0

Author: Done

```

You should see Alice_AEA running and showing its identity on the terminal. For example:

``` bash
My address is: YrP7H2qdCb3VyPwpQa53o39cWCDHhVcjwCtJLes6HKWM8FpVK
```

> Make note of this value. We will refer to this as Alice_AEA's address.

Reviewer: why?

Author: To enter in Faber_AEA's config

### Create Faber_AEA: 

In the fourth terminal, create a Faber_AEA and move into its project folder: 

``` bash
aea create faber
cd faber
```

### Add and Configure the Skill:

Add the `aries_faber` skill:

``` bash

> aea add --local skill fetchai/aries_faber:0.1.0

Reviewer: see above, no local

Author: Done

```

You then need to configure this skill. Open the skill's configuration file in `faber/vendor/fetchai/skills/aries_alice/skill.yaml` and ensure `admin_host` and `admin_port` details match those you noted above for Faber_ACA. In addition, make sure that the value of `alice_id` matches Alice_AEA's address as seen in the third terminal.

### Add and Configure the Connections:

Add `http_client`, `oef` and `webhook` connections:

``` bash

> aea add --local connection fetchai/http_client:0.1.0

Reviewer: no local

Author: Done. 

(I needed local at the time because of the local change I had to make to http_client)

aea add --local connection fetchai/webhook:0.1.0

> aea add connection fetchai/oef:0.1.0

Reviewer: 0.2.0

Author: Done.

Author: When adding the OEF 0.2.0 connection, I receive `Error: Not found in Registry.`

Reviewer: BEcause it will only be deployed with this.. So to test it atm you need `--local` flag

```

> You now need to configure the `webhook` connection. Open the connection's configuration file in `faber/vendor/fetchai/connections/webhook/connection.yaml` and ensure `webhook_port` matches with what you used above for Faber_ACA. 

Reviewer: see above

Author: Done

Next, make sure that the value of `webhook_url_path` is `/webhooks/topic/{topic}/`.

### Configure Faber_AEA:

> You now need to ensure that Faber_AEA uses the HTTP_Client connection as its default connection. Open the agent's configuration file in `faber/aea-config.yaml` and ensure that the `default_connection`'s value is `fetchai/http_client:0.1.0`.

Reviewer: see above

Author: Done
